Wake on LAN for UI-configured Samsung TVs

OnFreund opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

When configuring a Samsung TV through YAML, you can define an action to turn it on. This configuration is not available through the UI (for either manually initiated or through discovery).
Since I assume that in most cases the turn-on action would be wake on lan (and for more advanced cases one can use a [universal media player](https://www.home-assistant.io/integrations/universal/)), this PR adds the ability to set a MAC address through an options flow, which will be used to turn the TV on through wake on lan.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

OnFreund committed:
Samsung TV Options Flow skeleton

OnFreund commented:
@escoand this PR is just a placeholder and has no real functionality yet. If you think this is the right approach, I'll go ahead and complete it.

OnFreund commented:
@escoand let me know what you think. I'll happily complete this if you think this direction makes sense.

frenck commented:
@OnFreund Can you put the PR into "Draft" in the mean time?

There is a "Convert to draft" in the reviewers section in the sidebar of this PR.

escoand commented:
@OnFreund Interesting, never used such a options flow. How to use/test this?

OnFreund commented:
Once I implement this, it will be available as a gear button on the integration page. The demo integration has a functioning example that doesn't require special hardware.

escoand commented:
Nice, would love to see this. Best would be something like the automation editor UI (my TV is not able to wake-on-lan), but let's start with the MAC.

OnFreund commented:
The UI in options flow is very limited (same as config flow). We could potentially have a field for service name, and a field for json arguments, but there will be no autocompletion, and high possibility of misconfiguration.

OnFreund committed:
WoL with MAC address from options flow

OnFreund committed:
Test samsung options flow

OnFreund committed:
Test on-script creation from mac

OnFreund commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -174,3 +177,36 @@ def _try_connect(self):
+                current_mac = await self.hass.async_add_executor_job(
+                    partial(get_mac_address, **{"hostname": hostname})
+                )
+            except socket.gaierror:
```
for some reason when running the tests I'm getting `TypeError: catching classes that do not inherit from BaseException is not allowed` even though if I try this code in the console there's no problem.

MartinHjelmare replied:
Maybe just catch `OSError`?
```suggestion
            except OSError:
```

OnFreund commented:
@escoand this is the implementation I had in mind. There's one weird test failure that I can't understand, and a small section I'm not sure how to test.

OnFreund commented on homeassistant/components/samsungtv/media_player.py:
```diff
@@ -52,11 +56,27 @@
     on_script = None
-    if (
+
+    if CONF_MAC in config_entry.options:
```
Not sure how to test this part. If I understand correctly, this function is currently not covered by tests.

OnFreund replied:
@MartinHjelmare this is where coverage is impacted. It's definitely logic worth testing, but I don't believe this function is tested at all, and not sure how to go about it. Any help would be appreciated.

OnFreund replied:
@MartinHjelmare any guidance? thanks

escoand commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -174,3 +177,36 @@ def _try_connect(self):
+            except socket.gaierror:
+                current_mac = None
+
+        schema = {vol.Optional(CONF_MAC, default=current_mac): str}
```
Is there a cv type for mac address validation? Your current implementation accepts everything if I'm right.

OnFreund replied:
Unfortunately there isn't. According to https://github.com/home-assistant/architecture/issues/222 there also might never be.

escoand commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -174,3 +177,36 @@ def _try_connect(self):
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
```
No validation?

OnFreund replied:
The only risk here is a malformed MAC address, which is no different than a wrong MAC address. This is an advanced feature, and the input will most likely be pasted (unless obtained automatically). The wrong/malformed address will fail quickly and be easy to fix. Given all of the above, I don't think the validation overhead is worth it.
I've also quickly searched through the code base and couldn't find an integration that validates mac addresses.

escoand commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -174,3 +177,36 @@ def _try_connect(self):
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
```
How to remove the wake-on-lan config?

OnFreund replied:
good catch - I'll add this.

OnFreund committed:
Allow removal of mac address

OnFreund commented:
@escoand I think the request review button isn't working for this PR.
Mind taking another look?

escoand reviewed (approved):
Have not tested it, but it looks good in theory.

OnFreund commented:
Thanks. I tested it locally and it's working well.

I still don't understand that test failure, though. @MartinHjelmare any chance you can help with that?

OnFreund committed:
Fix handling of socket errors

OnFreund commented:
thanks @MartinHjelmare your suggestion fixed it.

@escoand in order to simulate this and test this case locally, I added a config entry with an invalid hostname. The logs are now full of:
```
2020-04-17 23:38:11 ERROR (MainThread) [homeassistant.helpers.entity] Update for media_player.dummy_samsung_tv fails
Traceback (most recent call last):
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/websocket/_http.py", line 143, in _get_addrinfo_list
    hostname, port, 0, 0, socket.SOL_TCP)
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/onfreund/dev/personal/home-assistant/homeassistant/helpers/entity.py", line 279, in async_update_ha_state
    await self.async_device_update()
  File "/Users/onfreund/dev/personal/home-assistant/homeassistant/helpers/entity.py", line 476, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/onfreund/dev/personal/home-assistant/homeassistant/components/samsungtv/media_player.py", line 140, in update
    self._state = STATE_ON if self._bridge.is_on() else STATE_OFF
  File "/Users/onfreund/dev/personal/home-assistant/homeassistant/components/samsungtv/bridge.py", line 66, in is_on
    return self._get_remote() is not None
  File "/Users/onfreund/dev/personal/home-assistant/homeassistant/components/samsungtv/bridge.py", line 256, in _get_remote
    self._remote.open()
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/samsungtvws/remote.py", line 146, in open
    sslopt=sslopt
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/websocket/_core.py", line 514, in create_connection
    websock.connect(url, **options)
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/websocket/_core.py", line 223, in connect
    options.pop('socket', None))
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/websocket/_http.py", line 113, in connect
    hostname, port, is_secure, proxy)
  File "/Users/onfreund/dev/personal/home-assistant/lib/python3.7/site-packages/websocket/_http.py", line 154, in _get_addrinfo_list
    raise WebSocketAddressException(e)
websocket._exceptions.WebSocketAddressException: [Errno 8] nodename nor servname provided, or not known
```
Might be worth seeing if this could be more gracefully handled.

OnFreund committed:
Upgrade getmac

escoand commented:
@OnFreund how did you add an invalid hostname? The config flow in the UI should fail. Have you edited the config entry?

OnFreund commented:
Yes, I edited the config entry. It's an edge case, but it could happen, if for example, the IP changed.

OnFreund committed:
Add tests for async_setup_entry

OnFreund committed:
Fix isort

Looks like the isort settings changed, so even tough this file
wasn't touched, it still needs changing.

OnFreund committed:
Use suggestd value instead of default

OnFreund committed:
Hanlde reloading of SamsungTV integration

OnFreund commented:
@escoand notice that I added the code needed to reload the integration. Would love for you to take another look. Thanks.

OnFreund commented:
The lint failures are coming from a different integration

OnFreund commented:
The failing test is in a different integration 🤷‍♂️

escoand commented on homeassistant/components/samsungtv/__init__.py:
```diff
@@ -62,8 +64,33 @@ async def async_setup(hass, config):
 
 async def async_setup_entry(hass, entry):
     """Set up the Samsung TV platform."""
+    undo_listener = entry.add_update_listener(_update_listener)
```
Do we need to unload and setup the entry again? The action is created at runtime and saved at `hass.data[DOMAIN][ip_address][CONF_ON_ACTION]`. Why do we not override it and use this directly in the entity?

OnFreund replied:
Interesting point. TBH I haven't thought of this direction until you wrote this.
Updating `hass.data[DOMAIN][ip_address][CONF_ON_ACTION]` won't be enough - the entity currently gets its processed `on_script` on initialization. We can have the entity listen to the update and handle accordingly. It might be a bit simpler, but potentially a bit more error prone.
The advantages of the unload/reload approach are:
1. Works through standard HA mechanisms.
2. Supports other unload scenarios. For example, removing a Samsung TV config entry currently suggests restarting HA. This wont' be needed if the integration can unload. Another example: if in the future we support editing the IP address (see https://github.com/home-assistant/architecture/issues/377), it will work out of the box.

Anyway, pros and cons to both approaches - I'm fine with whatever you choose.

escoand replied:
Are we going to wait for the linked issue? I would decide for the most future proof way.

OnFreund replied:
I don't see a reason to wait - this PR provides benefits as it is, even without the arch issue ever being implemented. The link is just to support the unload route as the more future proof one.

OnFreund commented:
Thanks @escoand.
@MartinHjelmare there were a few changes since your previous review. Would you mind taking a look?

MartinHjelmare commented:
I'll try to review this week.

elupus commented:
Alternate under consideration here: https://github.com/home-assistant/core/pull/35456

escoand commented:
As already said this PR is for me just a first step and something like the PR from @elupus makes me much more happy, especially because my TV doesn't support WOL but I can wake it with Kodi. So I would decide for the other instead of this. Sorry @OnFreund, but It's much more capable.

MartinHjelmare commented:
I think that device triggers is the way to go.

OnFreund commented:
OK, I converted this to draft for now, and will monitor the alternative. It does not yet have documentation, and I'm not sure I understand it (I've been using "classic" automations, not device automations, so I'm probably missing some context). Once I have a better grasp, I'll see if it makes sense here.

elupus commented:
It works with classic automations too. It just expose an event from the integration, which you hook up a turn on automation too, either manually or via device triggers

OnFreund commented:
@elupus cool. So I'm guessing device automation help abstract away the mapping to a specific device within the automation? And a classic automation can be a bit more generic, yet will have to work harder to map devices to actions?
The main question is how do you know if the media player entity supports being turned on?

As an aside (which is also relevant to the solution I proposed), also will be interesting in the future when a single device with multiple entities would need to be turned on. Which entity will be the "main" entity from a power perspective?

elupus commented:
You don't really know if it does support turn on sadly. 

It is theoretically possible to detect if there is any listeners registered on an event, and use that as basis for adding the support of turn on. Only problem would be that there would be in my case that there is one event for all entities (entity_id is in the event data), so you'd not really be able to detect if that specific entity support turn on. 

I've just written that off as a "oh well". The only real issue is from a gui perspective that it will show a turn on button which doesn't do anything.

OnFreund commented:
Thanks.

@escoand if we go with the automation approach, what we'd be missing is automatic detection of mac address (which I believe is the most common case by far).
Maybe we'll try to automatically detect Mac upon setup when we know the device is on (and if it's missing, try again when the state changes to on), expose it as an attribute, and add a wake-on-lan device action?

escoand commented:
@OnFreund I wouldn't do any kind of automatic mac usage or discovery. Just trigger the event on `turn_on` and let the user decide what to do.

OnFreund commented:
I would hate for an uneducated user to figure out the mac address of the device just to be able to turn it on, when we can easily get that in code.

@MartinHjelmare passing the ip address or host name in the event is not going to help much - when it's triggered the device is off, and resolution of host to mac will likely fail. That means that either the mac address, or a mapping from host to mac, will need to be available when the event is triggered.

escoand commented:
I don't have statistics over wol-able TVs but from the issues I had seen there are quite much which aren't. And from my point of view I don't like the idea of automatically created automations. Let's discover the mac address and push it along. But AFAIK this is impossible in some situations (e.g. VLAN), so we will end up with an unreliable solution.

elupus commented:
I think it would be better with a WOL switch entity, that will discover the Mac of a device on setup (config flow). That way the automation just need to trigger the switch.

escoand commented:
@elupus And whats the difference between the switch and the de-facto switch when adding the trigger? I would say the second is even better because you don't have an additional switch entity which is actually NOT a switch but just a trigger.

elupus commented:
Just the MAC address lookup. By a standard config entry, the MAC address can be deduced during config and stored. If unable to ARP find it, you can manually enter it.

WOL can work between VLAN if the broadcast domain of the subnet the other device is on is known.

However, if the MAC address is known by this integration by some other mean rather than ARP (sometimes exposed in metadata and/or SSDP discovery). Exposing hostname/ip as well as MAC in the event would allow hookup to a standard WOL service without a separate switch.

My point is really, i don't think this integration should try to deduce MAC, if the only reason is WOL.

OnFreund commented:
I think that users expect the ability to turn the device on to be supported without too much configuration. WOL is a common way to achieve that (probably *the* common way, by far, and also the one that's currently detailed in the docs), so I would think we should lower the barrier to using it, which means automatically obtaining the mac address.
Separate VLAN is a less interesting use case - if you're sophisticated enough to have HA and your TV on separate VLANs, you're a power user, and you can figure it out. My concern is the layman user.

escoand commented:
I was wondering how we could avoid all the "power button is not working" issues. What do you think of creating a default-deactivated automation and give a hint in the config flow that's maybe possible and the user has to try and activate/drop it?

The solution with the extra switch entity is the least preferred one.

OnFreund commented:
I like this (are there precedents of config flows creating automations?) as it reduces the friction to creating a functioning turn-on implementation in the common case. However, If I understand correctly, you would still have a "power button is not working" situation, it would just be a bit more explicit.

I'm not sure we can have the cake and eat it too - the event based solution has a lot of pros, but has a clear unmitigated con in the form of the power button not working unless setup. If we go with it, we need to live with the consequences.

escoand commented:
Yes, we could still have this issues, but when removing yaml some day, every user has the change to read this hint.

One other thing which comes to my mind: we have to setup wakeonlan as well.

OnFreund commented:
Seems like we have one solution that is generic and configurable (device trigger), and one that has a better user experience (`SUPPORT_TURN_ON` implemented correctly, automatic mac detection).
Before I give up, I want to make another proposal, which I believe combines the best of both worlds:
* The options flow will ask for the entity id of a script
* The existence of that entity id will turn on `SUPPORT_TURN_ON`
* When turning on, that script will be called with a parameter - the entity id of the media player
* In order to simply the common case of wake on lan, the config flow will detect the mac address (the device needs to be on during the config flow anyway), and the media player entity will expose it as an attribute. This also means that a single script can easily support multiple TVs
* For already existing config entries, a single attempt to obtain the mac will be made when the device switches state to on. If it fails, no more attempts will be made

This approach is fully generic and configurable, while also implementing `SUPPORT_TURN_ON` correctly, and simplifies the common wake on lan case. The only down side, is that it's slightly more complex than a device trigger (need to define a script with a parameter, and make sure you spell its entity id correctly in the options flow).
@escoand what do you think?

escoand commented:
Sorry for the delay, forgot to answer.

Your proposal is probably working, but as you already mentioned it's a bit a hacky for the user. Your only problem with the device trigger is the missing support flag, right? The UI is already able to list every use of an entity. Couldn't we use this discovery to add `SUPPORT_TURN_ON`?

escoand commented:
Could we additionally init a new automation edit screen as done from the device page and add a wake-on-lan call?

OnFreund commented:
> Your proposal is probably working, but as you already mentioned it's a bit a hacky for the user

I wouldn't call it "hacky". Yes, the initial setup is a bit more complicated, but ongoing functionality is better, I believe, for two reasons:
1. A single script can support multiple TVs. You can achieve that with device triggers as well, but I believe it'll be more complicated.
2. As you mentioned - `SUPPORT_TURN_ON` is implemented correctly with the script case. To implement correctly with an automation, you first have to find a way to list all of the uses (which, as you mentioned, is possible, since the UI does it), but then you have to figure out the actual use and whether it's for turning on, which is probably the complicated part.

OnFreund commented:
Anyway, I'm more than happy to tackle the "mac address as attribute" part. I'll defer to you on how to actually handle the turning on part.

OnFreund commented:
@escoand let me know if you'd like me to tackle the "mac address as attribute" part

escoand commented:
Yeah, do like you want. I haven't more or less to decide than any other. I've just added myself as codeowner to be informed not because I think the code is my own...

MartinHjelmare commented:
I suggest making an issue in our architecture repo first.

OnFreund commented:
Fair enough. I also have an idea how to make it more generic.
Will open an issue in the arch repo as soon as I can move forward with some of my other open PRs/Issues - trying to limit my WIP queue :)

OnFreund commented:
> Fair enough. I also have an idea how to make it more generic.

The idea I had in mind is actually very similar to the new automation variables in 0.115.
